### PR TITLE
Handle Python slice semantics in SimpleArray assignment

### LINKED ISSUE #856: SimpleArray should handle negative NumPy strides and Python slice semantics consistently
`SimpleArray` does not consistently follow Python/NumPy slicing semantics when negative bounds, negative steps, slice offsets, or negative NumPy strides are involved. Therefore, `SimpleArray` only works reliably when bounds, steps, and strides are positive and the array view starts from the beginning of the underlying buffer.

Affected APIs include:

- [`SimpleArray.__getitem__`](https://github.com/solvcon/modmesh/blob/ee8174b7b559ebef870ab63d0faf91a5c28584a8/cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp#L191-L194), implemented by [`SimpleArray::value_type const & at(ssize_t it)`](https://github.com/solvcon/modmesh/blob/ee8174b7b559ebef870ab63d0faf91a5c28584a8/cpp/modmesh/buffer/SimpleArray.hpp#L1802-L1813)
  - Example: `sarr[-8,-1,-2] = rhs`

- [`SimpleArray.__setitem__`](https://github.com/solvcon/modmesh/blob/master/cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp#L199), implemented by [`ArrayPropertyHelper::setitem_parser()`](https://github.com/solvcon/modmesh/blob/master/cpp/modmesh/buffer/pymod/array_common.hpp#L224-L305)
  - Example: `sarr[-8:-1:2, 8:1:-2] = rhs`

- [`SimpleArray(array=ndarray)` constructor](https://github.com/solvcon/modmesh/blob/master/cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp#L82-L149)
  - Example: `SimpleArrayFloat64(array=ndarr[::-1])`

## Tasks
This issues could be devided to a lot of tasks:
- [x] Support negetive index for [`SimpleArray.__getitem__`](https://github.com/solvcon/modmesh/blob/ee8174b7b559ebef870ab63d0faf91a5c28584a8/cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp#L191-L194). ([PR #875](https://github.com/solvcon/solvcon/pull/875))
- [x] Remove the usage of `SimpleArray::value_type const & at(size_t it)`. ([PR #879](https://github.com/solvcon/solvcon/pull/879))
- [x] Change stride storage from `small_vector<size_t>` to `small_vector<ssize_t>`. ([PR #883](https://github.com/solvcon/solvcon/pull/883))
- [x] Change shape and indices storage from `small_vector<size_t>` to `small_vector<ssize_t>`. ([PR #978](https://github.com/solvcon/solvcon/pull/978))
- [x] Fix [SimpleArray.__setitem__](https://github.com/solvcon/modmesh/blob/master/cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp#L199) by normalizing slice bounds, steps, and lengths with `pybind11::slice::compute()` and honoring the input NumPy strides, destination storage strides, slice starts, and slice steps. ([PR #857](https://github.com/solvcon/solvcon/pull/857))
- [x] Correct the buffer and logical-origin representation for NumPy views. The NumPy-array constructor currently uses the logical origin (`view_ptr`) as the start of `ConcreteBuffer`, so a negative-strided view omits valid elements before that pointer. The original view may still work because the complete NumPy allocation remains alive, but `clone()` copies only the incomplete range declared by `ConcreteBuffer` while retaining the original shape and strides. Reading the clone can then access memory before its allocation. See the [NumPy-array constructor](https://github.com/solvcon/solvcon/blob/master/cpp/solvcon/buffer/pymod/wrap_SimpleArray.hpp#L83-L126) and the [`SimpleArray` copy constructor](https://github.com/solvcon/solvcon/blob/master/cpp/solvcon/buffer/SimpleArray.hpp#L1767-L1773). ([PR #1101](https://github.com/solvcon/solvcon/pull/1101))
- [x] Support negative strides for `as_mdspan`, which includes: ([PR #1102](https://github.com/solvcon/solvcon/pull/1102))
  - [x] Add an internal signed-stride mdspan layout mapping. The current `std::layout_stride` mapping cannot safely represent negative strides and their logical-origin offset. ([PR #1102](https://github.com/solvcon/solvcon/pull/1102))
  - [x] Use the signed-stride mapping in `SimpleArray::as_mdspan()`. ([PR #1102](https://github.com/solvcon/solvcon/pull/1102))
- [x] Verify that the related issues continue to behave correctly. ([PR #875](https://github.com/solvcon/solvcon/pull/875), [PR #879](https://github.com/solvcon/solvcon/pull/879), [PR #883](https://github.com/solvcon/solvcon/pull/883), [PR #1094](https://github.com/solvcon/solvcon/pull/1094))

## Related Issues
* issue #580
* issue #603

## How to reproduce the error
```shell
$ python3 - <<'PY'
import numpy as np
import modmesh

ndarr_shape = (8, 9, 10, 11, 12)
ndarr = np.arange(np.prod(ndarr_shape), dtype='float64')
ndarr = ndarr.reshape(ndarr_shape)
expected = ndarr.copy()
sarr = modmesh.SimpleArrayFloat64(array=ndarr)

key = np.index_exp[-8:-1:2, 8:1:-2, 1:-1:3, ..., -2:-8:-2]
value_shape = expected[key].shape

value = np.empty(value_shape, dtype='float64')
next_value = 1000.0
for i0 in range(value_shape[0]):
    for i1 in range(value_shape[1]):
        for i2 in range(value_shape[2]):
            for i3 in range(value_shape[3]):
                for i4 in range(value_shape[4]):
                    value[i0, i1, i2, i3, i4] = next_value
                    next_value += 1

sarr[key] = value
expected[key] = value
np.testing.assert_array_equal(sarr.ndarray, expected)
PY
Traceback (most recent call last):
  File "<stdin>", line 23, in <module>
RuntimeError: Broadcast input array from shape(4, 4, 3, 11, 3) into shape(4, 4, 1, 11, 3)
```

```shell
$ python3 - <<'PY'
import numpy as np
import modmesh

ndarr_shape = (16, 18, 20, 22, 24)
ndarr = np.arange(np.prod(ndarr_shape), dtype='float64')
ndarr = ndarr.reshape(ndarr_shape)

view = ndarr[15::-2, 1::2, 19::-2, 1::2, 23::-2]
expected = np.array(view, copy=True)
sarr = modmesh.SimpleArrayFloat64(array=view)

key = np.index_exp[::-2, 1:-1:2, -8:-1:2, 10:2:-3, ...]
value_shape = expected[key].shape

value = np.empty(value_shape, dtype='float64')
next_value = 1000.0
for i0 in range(value_shape[0]):
    for i1 in range(value_shape[1]):
        for i2 in range(value_shape[2]):
            for i3 in range(value_shape[3]):
                for i4 in range(value_shape[4]):
                    value[i0, i1, i2, i3, i4] = next_value
                    next_value += 1

sarr[key] = value
expected[key] = value
np.testing.assert_array_equal(view, expected)
PY
Traceback (most recent call last):
  File "<stdin>", line 24, in <module>
RuntimeError: Broadcast input array from shape(4, 4, 4, 3, 12) into shape(18446744073709551612, 18446744073709551615, 4, 3, 12)
```

### How to reproduce the buffer-origin error

This example uses the same negative-strided NumPy view as PR #857.

The original `sarr` may still read correctly because the complete NumPy allocation remains alive, even though `ConcreteBuffer` omits elements before the view origin. `clone()` copies only that incomplete declared range while retaining the negative strides. Reading `clone.ndarray` then accesses memory before the cloned allocation and exits with status 139 (SIGSEGV).

```shell
$ python3 -u - <<'PY'
import numpy as np
import solvcon

ndarr_shape = (16, 18, 20, 22, 24)
ndarr = np.arange(np.prod(ndarr_shape), dtype='float64')
ndarr = ndarr.reshape(ndarr_shape)

view = ndarr[15::-2, 1::2, 19::-2, 1::2, 23::-2]
sarr = solvcon.SimpleArrayFloat64(array=view)

print("view shape:", view.shape)
print(
    "view strides:",
    tuple(stride // view.itemsize for stride in view.strides),
)
print("source sample:", sarr[1, 2, 3, 4, 5])

clone = sarr.clone()
print("clone created")

np.testing.assert_array_equal(clone.ndarray, view)
PY
view shape: (8, 9, 10, 11, 12)
view strides: (-380160, 21120, -1056, 48, -2)
source sample: 2530933.0
clone created
Segmentation fault (core dumped)
```

### PR DESCRIPTION
**Pull request by ThreeMonth03**
Fix `SimpleArray.__setitem__` by normalizing slices with `pybind11::slice::compute()`. The copy path now accounts for input NumPy strides, destination storage strides, slice starts, and slice steps.

## Previous failures

Negative bounds and steps could reject a valid assignment:

```python
key = np.index_exp[-8:-1:2, 8:1:-2, 1:-1:3, ..., -2:-8:-2]
sarr[key] = value
```

```text
RuntimeError: Broadcast input array from shape(4, 4, 3, 11, 3)
into shape(4, 4, 1, 11, 3)
```

A `SimpleArray` backed by a negative-strided NumPy view could produce unsigned-looking slice dimensions:

```python
view = ndarr[15::-2, 1::2, 19::-2, 1::2, 23::-2]
sarr = solvcon.SimpleArrayFloat64(array=view)
key = np.index_exp[::-2, 1:-1:2, -8:-1:2, 10:2:-3, ...]
sarr[key] = value
```

```text
RuntimeError: Broadcast input array from shape(4, 4, 4, 3, 12)
into shape(18446744073709551612, 18446744073709551615, 4, 3, 12)
```

Both assignments now update the same elements as NumPy.

Add regression coverage for negative-strided views, negative slice bounds and steps, zero steps, ellipsis handling, and excessive slices. Existing tests continue to cover assignment with ghost cells.

Related to #856.

### CODE REVIEW

**Comment by ThreeMonth03** [comment]
I notice that there are a lot of preparation works for this pull request, so I decide to postpone this pull request.

File: cpp/solvcon/buffer/pymod/array_common.hpp [hunk 1]
```diff
@@ -345,24 +345,37 @@ class ArrayPropertyHelper
...
+            throw pybind11::error_already_set();
+        }
 
-        slice_out[0] = start == "None" ? slice_out[0] : std::stoi(start);
-        slice_out[1] = stop == "None" ? slice_out[1] : std::stoi(stop);
-        slice_out[2] = step == "None" ? slice_out[2] : std::stoi(step);
+        slice_out[0] = start;
+        slice_out[1] = stop;
+        slice_out[2] = step;
+        slice_out[3] = slicelength;
```
Comment (ThreeMonth03) [inline cpp/solvcon/buffer/pymod/array_common.hpp:374]: Compute `start`, `stop`, `step` and `slicelength` by `pybind11::slice::compute`. It could handle negative `start`, `stop` and `step`.

File: cpp/solvcon/buffer/pymod/array_common.hpp [hunk 2]
```diff
@@ -388,7 +401,7 @@ class ArrayPropertyHelper
             }
         }
 
-        if (ellipsis_cnt + slice_cnt > ndim)
+        if (slice_cnt > ndim)
```
Comment (ThreeMonth03) [inline cpp/solvcon/buffer/pymod/array_common.hpp:400]: Ellipsis could be broadcasted to 0 dimension, so the condition of this runtime error should be fixed.
```python3
        sarr = solvcon.SimpleArrayFloat64((2, 3, 4))
        sarr[:, :, :, ...] = ndarr
        np.testing.assert_array_equal(sarr.ndarray, ndarr)
```

File: cpp/solvcon/buffer/pymod/TypeBroadcast.hpp [hunk 3]
```diff
@@ -16,12 +16,52 @@ namespace solvcon
...
+                                 sshape_type const & sidx)
+    {
+        ssize_t offset = 0;
+        for (ssize_t axis = 0; axis < arr_out.ndim(); ++axis)
+        {
+            ssize_t const index = slices[axis][0] + sidx[axis] * slices[axis][2];
+            offset += arr_out.stride(axis) * index;
+        }
+        return offset;
+    }
```
Comment (ThreeMonth03) [inline cpp/solvcon/buffer/pymod/TypeBroadcast.hpp:63]: Wrap the repetitive logics as helper functions.

File: cpp/solvcon/buffer/pymod/TypeBroadcast.hpp [hunk 4]
```diff
@@ -34,32 +74,8 @@ struct TypeBroadcastImpl
...
-            ssize_t offset_out = 0;
-            ssize_t const ndim = arr_out.ndim();
-            for (ssize_t it = 0; it < ndim; ++it)
-            {
-                auto const axis_index = static_cast<size_t>(it);
-                ssize_t const step = slices[axis_index][2];
-                offset_out += arr_out.stride(it) * sidx[axis_index] * step;
-            }
+            const D * ptr_in = arr_in->data() + input_offset(*arr_in, sidx);
+            ssize_t const offset_out = output_offset(arr_out, slices, sidx);
```
Comment (ThreeMonth03) [inline cpp/solvcon/buffer/pymod/TypeBroadcast.hpp:78]: Simplify depth-first search logic, the program doesn't enumerate indices and assign value simultaneously now.

File: tests/test_buffer.py [hunk 5]
```diff
@@ -1086,6 +1117,30 @@ def test_SimpleArray_broadcast_slice_ghost_md(self):
...
+        key = np.index_exp[-8:-1:2, 8:1:-2, 1:-1:3, ..., -2:-8:-2]
+        value_shape = expected[key].shape
+        value = np.arange(np.prod(value_shape), dtype='float64')
+        value = value.reshape(value_shape) + 1000.0
+
+        sarr[key] = value
+        expected[key] = value
+
+        np.testing.assert_array_equal(sarr.ndarray, expected)
+        np.testing.assert_array_equal(ndarr, expected)
```
Comment (ThreeMonth03) [inline tests/test_buffer.py:1136]: Check whether SimpleArray can deal with negative `start` and `end` now.

File: tests/test_buffer.py [hunk 6]
```diff
@@ -1086,6 +1117,30 @@ def test_SimpleArray_broadcast_slice_ghost_md(self):
...
+
+    def test_SimpleArray_broadcast_slice_zero_step(self):
+        ndarr = np.arange(4 * 5 * 6, dtype='float64').reshape((4, 5, 6))
+        sarr = solvcon.SimpleArrayFloat64(array=ndarr)
+        with self.assertRaisesRegex(ValueError, "slice step cannot be zero"):
+            sarr[::0, :, :] = np.zeros((4, 5, 6), dtype='float64')
```
Comment (ThreeMonth03) [inline tests/test_buffer.py:1142]: Check whether SimpleArray can raise error for 0 step now.

File: tests/test_buffer.py [hunk 7]
```diff
@@ -1047,6 +1047,37 @@ def test_SimpleArray_broadcast_slice_shape(self):
...
+
+        key = np.index_exp[::-2, 1:-1:2, -8:-1:2, 10:2:-3, ...]
+        value_shape = expected[key].shape
+        value = np.arange(np.prod(value_shape), dtype='float64')
+        value = value.reshape(value_shape) + 1000.0
+
+        sarr[key] = value
+        expected[key] = value
+
+        np.testing.assert_array_equal(sliced_ndarr, expected)
```
Comment (ThreeMonth03) [inline tests/test_buffer.py:1079]: Check whether SimpleArray can deal with negative steps now.

File: tests/test_buffer.py [hunk 8]
```diff
@@ -1047,6 +1047,37 @@ def test_SimpleArray_broadcast_slice_shape(self):
         ):
             sarr[::2, ::1, ...] = ndarr[...]
 
+        sarr = solvcon.SimpleArrayFloat64((2, 3, 4))
+        sarr[:, :, :, ...] = ndarr
+        np.testing.assert_array_equal(sarr.ndarray, ndarr)
```
Comment (ThreeMonth03) [inline tests/test_buffer.py:1052]: Add test for extending eclipsis to zero dimension.

File: tests/test_buffer.py [hunk 9]
```diff
@@ -1047,6 +1047,37 @@ def test_SimpleArray_broadcast_slice_shape(self):
...
+
+        with self.assertRaisesRegex(
+                RuntimeError,
+                r"syntax error\. dimensions mismatches"
+        ):
+            sarr[:, :, :, :] = np.zeros((2, 3, 4, 1), dtype='float64')
```
Comment (ThreeMonth03) [inline tests/test_buffer.py:1058]: Add test for dimensions mismatches.

**Review by ThreeMonth03 — COMMENTED** [review]
@yungyuc Please take a look.

File: cpp/solvcon/buffer/pymod/array_common.hpp [hunk 10]
```diff
@@ -345,24 +345,37 @@ class ArrayPropertyHelper
...
+                           ssize_t length)
     {
-        auto start = std::string(pybind11::str(slice_in.attr("start")));
-        auto stop = std::string(pybind11::str(slice_in.attr("stop")));
-        auto step = std::string(pybind11::str(slice_in.attr("step")));
+        pybind11::ssize_t start = 0;
+        pybind11::ssize_t stop = 0;
+        pybind11::ssize_t step = 0;
+        pybind11::ssize_t slicelength = 0;
+        if (!slice_in.compute(length,
```
Comment (yungyuc) [inline cpp/solvcon/buffer/pymod/array_common.hpp]: Use a single line for the call to `compute()`.

File: cpp/solvcon/buffer/pymod/array_common.hpp (hunk 2 above)
Comment (yungyuc) [reply cpp/solvcon/buffer/pymod/array_common.hpp:400]: Good catch.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Slicing and indexing higher dimension are hard.  Thanks for taking care of this.  This is a good addition.

Please fix the formatting issue:
- [x] Use a single line for the call to `compute()`.

File: cpp/solvcon/buffer/pymod/array_common.hpp [hunk 11]
```diff
@@ -345,24 +345,33 @@ class ArrayPropertyHelper
...
+                           ssize_t length)
     {
-        auto start = std::string(pybind11::str(slice_in.attr("start")));
-        auto stop = std::string(pybind11::str(slice_in.attr("stop")));
-        auto step = std::string(pybind11::str(slice_in.attr("step")));
+        pybind11::ssize_t start = 0;
+        pybind11::ssize_t stop = 0;
+        pybind11::ssize_t step = 0;
+        pybind11::ssize_t slicelength = 0;
+        if (!slice_in.compute(length, &start, &stop, &step, &slicelength))
```
Comment (ThreeMonth03) [inline cpp/solvcon/buffer/pymod/array_common.hpp:366]: Use a single line for the call to compute().

**Review by ThreeMonth03 — COMMENTED** [review]
- [x] Use a single line for the call to compute().
@yungyuc Thanks for your review. Please review the modification again.

**Review by yungyuc — APPROVED** [review]
LGTM.  Nice patch.
